Fix incorrect intent count metric

mrtracy opened the pull request:
The displayed metric for "intent count" on a store was being computed
incorrectly in the event of a split.

The issue was that the normal incremental stats computed during command
execution are zeroed out when an EndTransactionRequest has a splitTrigger - the
expectation is that the MVCC stats will be explicitly recomputed for both
ranges, and thus the incremental update is not needed. However, the splitTrigger
is thus responsible for updating the store metrics, and it was not doing this.

Fixes #2952

<!-- Reviewable:start -->

[<img src="https://reviewable.io/review_button.svg" height="40" alt="Review on Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/4624)

<!-- Reviewable:end -->

Commit:
Fix incorrect intent count metric

The displayed metric for "intent count" on a store was being computed
incorrectly in the event of a split.

The issue was that the normal incremental stats computed during command
execution are zeroed out when an EndTransactionRequest has a splitTrigger - the
expectation is that the MVCC stats will be explicitly recomputed for both
ranges, and thus the incremental update is not needed. However, the splitTrigger
is thus responsible for updating the store metrics, and it was not doing this.

Commit also adds a basic test to ensure that metric counters are accurate on the
store after a split has occurred.

Fixes #2952

tamird commented:
LGTM but needs a test.

---

Reviewed 1 of 1 files at r1.
Review status: all files reviewed at latest revision, all discussions resolved, some commit checks failed.

---



*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/4624#-:-KBKolEHyE53NAOYVgF8:-575562028)*
<!-- Sent from Reviewable.io -->

petermattis commented:
Heh, I had written almost exactly the same thing as @tamird. Connectivity on the train caused my message to not be sent.

tschottdorf commented:
@nvanbenschoten you should greenlight this as you're not the resident `MVCCStats` expert. Sorry.

nvanbenschoten commented:
LGTM, although I have a PR on the way that rips a lot of this code up. If my change goes in after this, which I assume it will, I will make sure that this fix still works. Like Tamir mentioned, a test would help guarantee that.

Commit:
Bump max byte counts for TestNodeStatus

The TestNodeStatus CLI test attempts to enforce a sort of 'sanity check' on the
metrics it returns, ensuring that they are between a min and max value. Due to
the asynchronous nature of this test, it is impossible to expect an exact
number.

Recent changes ensure that the "system_bytes" value will occassionally be above
the current maximum of 20000, causing the test to fail. I have boosted all maximum
"bytes" values to 30000 in this commit.

This is the second time these maximum values have needed a boost; we may want to
reconsider the value of this check if they need to be boosted again in the
future.

mrtracy commented:
Added a test to verify stats after a split. That was easier than I thought - I was certain this sort of verification was going to be incredibly flaky, but it seems quite reliable.

Also fixed another test "TestNodeStatus", which performs a range-based sanity check - we have apparently exceeded that range (this is something that has occurred in the past as well).

This also takes us part of the way to #4583, but we should also be verifying statistics in a few more scenarios (not just splits).

tamird commented:
LGTM. Thanks for adding the test.

---

Reviewed 2 of 2 files at r2, 1 of 1 files at r3.
Review status: all files reviewed at latest revision, 4 unresolved discussions, some commit checks failed.

---

*[storage/client_metrics_test.go, line 1 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4624#-KBOsl1xufmpcYvEL-RZ:-KBOsl1xufmpcYvEL-R_:676036456) ([raw file](https://github.com/cockroachdb/cockroach/blob/8681228396ca9304474115f7d75dbcde03e79cdf/storage/client_metrics_test.go#L1)):*
2016

---

*[storage/client_metrics_test.go, line 28 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4624#-KBOt7wEZNFPFc70A45m:-KBOt7wEZNFPFc70A45n:-1171089443) ([raw file](https://github.com/cockroachdb/cockroach/blob/8681228396ca9304474115f7d75dbcde03e79cdf/storage/client_metrics_test.go#L28)):*
The line numbers that these `Fatalf`s and `Errorf`s will produce will be quite frustrating for anyone looking at this test. Either use `util.ErrorfSkipFrames` to generate the error message, or return an error from this method and call `Fatal` at the caller.

Also, the `return 0`s are not necessary.

---

*[storage/client_metrics_test.go, line 87 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4624#-KBOtbniq3Y4y-iwfO6-:-KBOtbniq3Y4y-iwfO60:-1759607936) ([raw file](https://github.com/cockroachdb/cockroach/blob/8681228396ca9304474115f7d75dbcde03e79cdf/storage/client_metrics_test.go#L87)):*
s/<=/</

---

*[storage/store.go, line 2019 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4624#-KBOscAbwzu_nHXyPwSP:-KBOscAbwzu_nHXyPwSQ:2004952960) ([raw file](https://github.com/cockroachdb/cockroach/blob/8681228396ca9304474115f7d75dbcde03e79cdf/storage/store.go#L2019)):*
nit: `s.Clock().PhysicalNow()`

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/4624#-:-KBOtfgr6Bl269BL7h7-:1789412593)*
<!-- Sent from Reviewable.io -->

mrtracy commented:
Review status: 0 of 4 files reviewed at latest revision, 4 unresolved discussions.

---

*[storage/client_metrics_test.go, line 1 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4624#-KBOsl1xufmpcYvEL-RZ:-KBPFHI8cp4JTZUlG23j:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/8681228396ca9304474115f7d75dbcde03e79cdf/storage/client_metrics_test.go#L1)):*
Done.

---

*[storage/client_metrics_test.go, line 28 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4624#-KBOt7wEZNFPFc70A45m:-KBPF2TZJ9ZRafgB-1bv:334198410) ([raw file](https://github.com/cockroachdb/cockroach/blob/8681228396ca9304474115f7d75dbcde03e79cdf/storage/client_metrics_test.go#L28)):*
Done, but I don't believe it was going to be very frustrating; each invocation of "checkCounters" uses a unique key, and the key is present in the error message.

---

*[storage/client_metrics_test.go, line 87 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4624#-KBOtbniq3Y4y-iwfO6-:-KBPFTLMm060Bnohw8lf:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/8681228396ca9304474115f7d75dbcde03e79cdf/storage/client_metrics_test.go#L87)):*
Done.

---

*[storage/store.go, line 2019 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4624#-KBOscAbwzu_nHXyPwSP:-KBPFiCkp1A7kkteBCA-:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/8681228396ca9304474115f7d75dbcde03e79cdf/storage/store.go#L2019)):*
Done.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/4624)*
<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 4 of 4 files at r4.
Review status: all files reviewed at latest revision, all discussions resolved, some commit checks failed.

---



*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/4624)*
<!-- Sent from Reviewable.io -->
